app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+   * @param onSuccess To be called once the details have been successfully retrieved. Do not use
+   *   this to retrieve the details values, rather observe the hike in [hikeFlows] to be notified
+   *   automatically.
+   */
+  fun computeDetailsFor(hikeId: String, onSuccess: () -> Unit = {}, onFailure: () -> Unit = {}) =
+      viewModelScope.launch { computeDetailsForAsync(hikeId, onSuccess, onFailure) }
+
+  private suspend fun selectHikeAsync(
+      hikeId: String,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ) =
+      withContext(dispatcher) {
+        var successful = false
+        _hikesMutex.withLock {
+          // Retrieve the hike to select from the loaded hikes
+          val hike = _hikeFlowsMap[hikeId] ?: return@withLock
+
+          // Mark the hike as selected
+          _selectedHikeId = hikeId
+          _selectedHike.value = hike.value
+          successful = true
+        }
+
+        // Perform the callback without the lock, to avoid deadlocks and improve performance
+        if (successful) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  private suspend fun unselectHikeAsync() =
+      _hikesMutex.withLock {
+        // Only emit null as a value if the selected hike was not already null
+        if (_selectedHikeId != null) {
+          _selectedHikeId = null
+          _selectedHike.value = null
+        }
+      }
+
+  /**
+   * This function performs a loading operation but DOES NOT set [_loading]. If calling it from
+   * another function, please set [_loading] to true before calling the function and to false once
+   * the call has ended.
+   *
+   * @return True if the operation is successful, false otherwise.
+   */
+  private suspend fun refreshSavedHikesCacheAsync(
+      onSuccess: () -> Unit = {},
+      onFailure: () -> Unit = {}
+  ): Boolean =
+      withContext(dispatcher) {
+        val savedHikes: List<SavedHike>
+        try {
+          // Load the saved hikes from the repository
+          savedHikes = savedHikesRepo.loadSavedHikes()
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error encountered while loading saved hikes", e)
+          onFailure()
+          return@withContext false
+        }
+
+        // Wait for the lock to avoid concurrent modifications
+        _hikesMutex.withLock { updateSavedHikesCache(savedHikes) }
+
+        onSuccess()
+        return@withContext true
+      }
+
+  /**
+   * Helper function for [refreshSavedHikesCacheAsync].
+   *
+   * Takes a list of saved hikes and updates the local cache of saved hikes with that list.
+   *
+   * It is the caller's responsibility to ensure the saved hikes list is up-to-date and to acquire
+   * [_hikesMutex] before calling this function.
+   *
+   * @param newList The list of saved hikes to update the cache with.
+   */
+  private fun updateSavedHikesCache(newList: List<SavedHike>) {
```

Comment:
I rewrote the function to completely separate the cases where `_loadedHikesType == LoadedHikes.FromSaved` from `_loadedHikesType == LoadedHikes.FromBounds`, to make it clearer.

So I guess you can say it "returns early" now, because of the condition.

Reply:
Mmmmh, I see, but now this code is duplicated. How about just having a check, (really) return early and deduplicate this code?

```kt
      _hikeFlowsMap.forEach { (hikeId, hikeFlow) ->
        val savedHike = _savedHikesMap[hikeId]
        val hike = hikeFlow.value
        val (changeNeeded, updated) = hikeNeedsSavedStatusUpdate(hike, savedHike)
        if (changeNeeded) {
          hikeFlow.value = updated
        }
      }
```

Reply:
I introduced an early return clause if the no hikes are loaded and extracted the duplicated code to a helper function 👌🏻

Reply:
Thanks for your effort!